### PR TITLE
Support lifo eviction on ctx cancel

### PR DESCRIPTION
**Pull request by marcoferrer**
One issue Im running into is a case where a context can become cancelled before the max backlog timeout has a chance to elapse. This proposed change introduces a configuration field to enable backlog eviction if the context used to acquire a token ends up cancelled. To support adding a new configuration for the lifo blocking limiter constructor without breaking existing users I added a new constructor that takes a configuration struct. This was the best solution I could come up with that would prevent introducing breaking changes.

Please let me know if support for this use case is outside of the scope of this library and I can close accordingly.

### CODE REVIEW

File: limiter/lifo_blocking.go [hunk 1]
```diff
@@ -11,7 +11,7 @@ import (
 
 type lifoElement struct {
 	ctx         context.Context
-	releaseChan chan core.Listener
+	releaseChan chan<- core.Listener
```
Comment (marcoferrer) [inline limiter/lifo_blocking.go:14]: dumb nit, restricting channel types

File: limiter/lifo_blocking.go [hunk 2]
```diff
@@ -173,56 +173,94 @@ func (l *LifoBlockingListener) OnSuccess() {
...
 
-	registry.RegisterGauge(core.MetricLifoQueueSize, core.NewUint64MetricSupplierWrapper(l.backlog.len), tags...)
+	config.MetricRegistry.RegisterGauge(
+		core.MetricLifoQueueSize, core.NewUint64MetricSupplierWrapper(l.backlog.len), config.Tags...)
 
 	return l
 }
 
+// NewLifoBlockingLimiter will create a new LifoBlockingLimiter
+// Deprecated, use NewLifoBlockingLimiterFromConfig instead
```
Comment (marcoferrer) [inline limiter/lifo_blocking.go:237]: I kept the original constructor and delegated it to the new one

File: limiter/lifo_blocking.go [hunk 3]
```diff
@@ -252,6 +300,12 @@ func (l *LifoBlockingLimiter) tryAcquire(ctx context.Context) core.Listener {
 		// Remove the holder from the backlog.
 		evict()
 		return nil
+	case <-ctxDone:
```
Comment (marcoferrer) [inline limiter/lifo_blocking.go:303]: If `backlogEvictDoneCtx` is false, `ctxDone` will end up nil and block forever. Preserving existing behavior

**Review by platinummonkey — APPROVED** [review]
Nice! Thanks for the details!

**Pull request merged**